syntax = "proto3";

//blockchain service and the function definition
service BlockChain {
  rpc initTxList(InitTxListRequest) returns (InitTxListResponse) {}
  rpc addNewBlock(AddBlockRequest) returns (AddBlockResponse) {}
  rpc receiveBlock(ReceiveBlockRequest) returns (ReceiveBlockResponse) {}
  rpc receiveMessage(receiveMessageRequest) returns (receiveMessageResponse) {}
  rpc getUTXOs(getUTXOsRequest) returns (getUTXOsResponse){}
  rpc getState(getStateRequest) returns (getStateResponse) {}
  rpc QueryDB(QueryDBRequest) returns (QueryDBResponse) {}
  rpc addNewTransaction(addNewTransactionRequest) returns (addNewTransactionResponse) {}
  rpc sendTransaction(sendTransactionRequest) returns (sendTransactionResponse) {}
}

message InitTxListRequest{
  string message = 1;
}

message InitTxListResponse{
  string message = 1;
}

message AddBlockRequest {
  string transaction = 1;
  int32 nonce = 2;
}

message AddBlockResponse {
  string hash = 1;
  Block newBlock = 2;
}

//block structure definition
message Block {

  int32 index = 1;
  string hash = 2;
  string prevBlockHash = 3 ;
  string rootHash = 4;
  int32 difficulty = 5;
  int32 nonce = 6;
  string timestamp = 7;

  repeated Transaction transactionList = 8;
}

message Transaction {
  string TxId = 1;
  repeated TxIn TxInList = 2;
  repeated TxOut TxOutList = 3;

}

message TxIn {
  string TxOutId = 1;
  int32 TxOutIndex = 2;
  string signature = 3;
}

message TxOut {
  string address = 1;
  double amount = 2;
}

message ReceiveBlockRequest{
  string message = 1;
  Block newBlock = 2;
}

message ReceiveBlockResponse{
  string message = 1;
}

message UTXOs {
  repeated string key = 1;
  repeated double amount = 2;
  repeated string owner = 3;
}

message getUTXOsRequest{
  string message = 1;

}

message getUTXOsResponse{
  string message = 1;
  UTXOs utxos = 2;
}

message getStateRequest{
  int32 message = 1;
}

message getStateResponse{
  string message = 1;
}

message addNewTransactionRequest {
  Transaction tx = 1;
}

message addNewTransactionResponse {
  string message = 1;
}

message sendTransactionRequest {
  Transaction tx = 1;
}

message sendTransactionResponse {
  string message = 1;
}

message QueryDBRequest{
  string message = 1;
}

message QueryDBResponse{
  string message = 1;
}

message QueryLatestBlockIdxRequest{
  string message = 1;
}

message QueryLatestBlockIdxResponse {
  int32 index = 1;
}

message receiveMessageRequest{
  string message = 1;
}

message receiveMessageResponse {
  string message = 1;
  Block newBlock = 2;
}